contracts/Voting.sol

```diff
@@ -1,99 +1,213 @@
         voteVerifier = voteVerifier_;
-        registerVerifier = registerVerifier_;
+        registerVerifier = IRegisterVerifier(registerVerifier_);
+
+        smtTreeMaxDepth = treeHeight_;
     }
 
+    /**
+     * @inheritdoc IVoting
+     */
+    function __Voting_init(VotingParams calldata votingParams_) external initializer {
+        __Ownable_init();
+        __PoseidonSMT_init(smtTreeMaxDepth);
+
+        _validateVotingParams(votingParams_);
+
+        votingInfo.remark = votingParams_.remark;
+        votingInfo.values.commitmentStartTime = votingParams_.commitmentStart;
+        votingInfo.values.votingStartTime =
+            votingParams_.commitmentStart +
+            votingParams_.commitmentPeriod;
+        votingInfo.values.votingEndTime =
+            votingInfo.values.votingStartTime +
+            votingParams_.votingPeriod;
+
+        for (uint256 i = 0; i < votingParams_.candidates.length; i++) {
+            candidates[votingParams_.candidates[i]] = true;
+        }
+
+        emit VotingInitialized(msg.sender, votingParams_);
+    }
+
+    /**
+     * @inheritdoc IVoting
+     */
     function registerForVoting(
-        bytes32 commitment_,
-        VerifierHelper.ProofPoints calldata proof_
+        IBaseVerifier.ProveIdentityParams memory proveIdentityParams_,
+        IRegisterVerifier.RegisterProofParams memory registerProofParams_,
+        IBaseVerifier.TransitStateParams memory transitStateParams_,
+        bool isTransitState_
     ) external {
-        require(!commitments[commitment_], "Voting: commitment already exists");
-
         require(
-            registerVerifier.verifyProofSafe([uint256(commitment_)].asDynamic(), proof_, 1),
-            "Voting: Invalid vote proof"
+            getProposalStatus() == VotingStatus.COMMITMENT,
+            "Voting: the voting must be in the commitment state to register"
         );
 
+        bytes32 commitment_ = registerProofParams_.commitment;
+
+        require(!commitments[commitment_], "Voting: commitment already exists");
+
+        IRegisterVerifier.RegisterProofInfo memory registerProofInfo_ = IRegisterVerifier
+            .RegisterProofInfo({
+                registerProofParams: registerProofParams_,
+                votingAddress: address(this)
+            });
+
+        if (isTransitState_) {
+            registerVerifier.transitStateAndProveRegistration(
+                proveIdentityParams_,
+                registerProofInfo_,
+                transitStateParams_
+            );
+        } else {
+            registerVerifier.proveRegistration(proveIdentityParams_, registerProofInfo_);
+        }
+
         _add(commitment_);
         commitments[commitment_] = true;
         rootsHistory[getRoot()] = true;
 
-        emit UserRegistered(commitment_, block.number);
+        emit UserRegistered(msg.sender, proveIdentityParams_, registerProofParams_, block.number);
     }
 
+    /**
+     * @inheritdoc IVoting
+     */
     function vote(
```

Comment:
What is the relationship between commitment and nullifier? I see no circom circuits here

Reply:
Same as always: 
H(Secret | Nullifier)  = Commitment 
H(Nullifier) = Nullifier Hash

Repository with relevant circuits: https://github.com/rarimo/passport-zk-circuits